### DISCUSSION TITLE
[Feature] Support MySQL help command in OceanBase

### DISCUSSION THREAD

**Discussion by Junkrat77** [oceanbase/oceanbase#992] [Classroom]
<div class="lake-content" typography="classic"><h1 id="ZYNmM" style="font-size: 28px; line-height: 36px; margin: 26px 0 10px 0"><span class="ne-text">Abstract</span></h1><p id="u50d57e12" class="ne-p" style="margin: 0; padding: 0; min-height: 24px"><span class="ne-text">In order to improve the experience of open-source OceanBase(OB), we think OB needs to support command line help syntax just like MySQL.</span></p><p id="ua07d0b37" class="ne-p" style="margin: 0; padding: 0; min-height: 24px"><span class="ne-text">For examle:</span></p><p id="u1aadd5c3" class="ne-p" style="margin: 0; padding: 0; min-height: 24px"><img src="https://intranetproxy.alipay.com/skylark/lark/0/2022/png/65656285/1659064048811-04cb8051-2cad-4ecf-8fd7-3dbf09cd37fd.png" width="453.5" id="u2c9b51c6" class="ne-image"><img src="https://intranetproxy.alipay.com/skylark/lark/0/2022/png/65656285/1659064006894-2c04f4af-0bcc-4cae-845f-fdf1b2e3462e.png" width="456.5" id="u81b39ee2" class="ne-image"></p><p id="u065ba5b9" class="ne-p" style="margin: 0; padding: 0; min-height: 24px"><br></p><h1 id="Tvi0C" style="font-size: 28px; line-height: 36px; margin: 26px 0 10px 0"><span class="ne-text">Design</span></h1><p id="udc2087bc" class="ne-p" style="margin: 0; padding: 0; min-height: 24px"><span class="ne-text">In order to make Oceanbase support 'help' command, we need to create corresponding help tables, parser the 'help' command. After parsing it, we search the help tables and return corresponding contents to client.</span></p>
<h2 id="k62Ha" style="font-size: 24px; line-height: 32px; margin: 21px 0 5px 0"><span class="ne-text">Help Tables</span></h2><p id="u685b2318" class="ne-p" style="margin: 0; padding: 0; min-height: 24px"><span class="ne-text">Just like MySQL, we have defined four help tables under the </span><strong><span class="ne-text">mysql database</span></strong><span class="ne-text"> in OB, and they are all inner system table which would be created and initialized during cluster bootstrapping and defined as follows:</span></p>
<p id="uf66b75cd" class="ne-p" style="margin: 0; padding: 0; min-height: 24px"><strong><span class="ne-text">Help_category</span></strong></p><p id="u7b5cb435" class="ne-p" style="margin: 0; padding: 0; min-height: 24px; text-indent: 2em"><span class="ne-text">Description: table holding system help categories which is the top-level concept.</span></p><p id="u76bc51a1" class="ne-p" style="margin: 0; padding: 0; min-height: 24px; text-indent: 2em"><span class="ne-text" style="color: rgb(36, 41, 47); font-size: 14px">Struction</span><strong><span class="ne-text" style="color: rgb(36, 41, 47); font-size: 14px">:</span></strong></p>

Column | Type | Description
-- | -- | --
help_category_id | int | category id
name | varchar(64) | category name
parent_category_id | int | parent category id
url | text | the url of the category

<p id="u8b09edc5" class="ne-p" style="margin: 0; padding: 0; min-height: 24px; text-indent: 2em"><span class="ne-text">Definition:</span></p>
<pre>def_table_schema(
  database_id    = 'OB_MYSQL_SCHEMA_ID',
  table_name    = 'help_relation',
  table_id      = '423',
  table_type = 'SYSTEM_TABLE',
  gm_columns = [],
  rowkey_columns = [
      ('help_topic_id', 'int'),
      ('help_keyword_id', 'int')
  ],
  in_tenant_space = True,
  normal_columns = [],
  columns_with_tenant_id = [],
)      
</pre>

<p id="uf66b75cd" class="ne-p" style="margin: 0; padding: 0; min-height: 24px"><strong><span class="ne-text">Help_topic</span></strong></p><p id="u7b5cb435" class="ne-p" style="margin: 0; padding: 0; min-height: 24px; text-indent: 2em"><span class="ne-text">Description: table holding system help topic which is the usage of sql statement and function, or the introduction of a feature.</span></p><p id="u76bc51a1" class="ne-p" style="margin: 0; padding: 0; min-height: 24px; text-indent: 2em"><span class="ne-text" style="color: rgb(36, 41, 47); font-size: 14px">Struction</span><strong><span class="ne-text" style="color: rgb(36, 41, 47); font-size: 14px">:</span></strong></p>

Column | Type | Description
-- | -- | --
help_topic_id | int | topic id
name | varchar(64) | topic name
help_category_id | int | corresponding policy id
description | text | the description of the topic
example | text | example of how to use
url | text | link url

<p id="u8b09edc5" class="ne-p" style="margin: 0; padding: 0; min-height: 24px; text-indent: 2em"><span class="ne-text">Definition:</span></p>
<pre>def_table_schema(
  database_id    = 'OB_MYSQL_SCHEMA_ID',
  table_name    = 'help_topic',
  table_id      = '422',
  table_type = 'SYSTEM_TABLE',
  gm_columns = [],
  rowkey_columns = [
      ('help_topic_id', 'int'),
  ],
  in_tenant_space = True,
  normal_columns = [
    ('name', 'varchar:64'),
    ('help_category_id', 'int'),
    ('description', 'text'),
    ('example', 'text'),
    ('url', 'text')
  ],
  columns_with_tenant_id = [],
)
</pre>

<p id="uf66b75cd" class="ne-p" style="margin: 0; padding: 0; min-height: 24px"><strong><span class="ne-text">Help_keyword</span></strong></p><p id="u7b5cb435" class="ne-p" style="margin: 0; padding: 0; min-height: 24px; text-indent: 2em"><span class="ne-text">Description: table holding some keywords. Every keyword is related to some topics.</span></p><p id="u76bc51a1" class="ne-p" style="margin: 0; padding: 0; min-height: 24px; text-indent: 2em"><span class="ne-text" style="color: rgb(36, 41, 47); font-size: 14px">Struction</span><strong><span class="ne-text" style="color: rgb(36, 41, 47); font-size: 14px">:</span></strong></p>

Column | Type | Description
-- | -- | --
help_keyword_id | int | keyword id
name | varchar(64) | keyword name

<p id="u8b09edc5" class="ne-p" style="margin: 0; padding: 0; min-height: 24px; text-indent: 2em"><span class="ne-text">Definition:</span></p>
<pre>def_table_schema(
  database_id    = 'OB_MYSQL_SCHEMA_ID',
  table_name    = 'help_keyword',
  table_id      = '421',
  table_type = 'SYSTEM_TABLE',
  gm_columns = [],
  rowkey_columns = [
      ('help_keyword_id', 'int'),
  ],
  in_tenant_space = True,
  normal_columns = [
    ('name', 'varchar:64')
  ],
  columns_with_tenant_id = [],
)
</pre>

<p id="uf66b75cd" class="ne-p" style="margin: 0; padding: 0; min-height: 24px"><strong><span class="ne-text">Help_relation</span></strong></p><p id="u7b5cb435" class="ne-p" style="margin: 0; padding: 0; min-height: 24px; text-indent: 2em"><span class="ne-text">Description: Every table item represents a pair of keyword and topic associated.</span></p><p id="u76bc51a1" class="ne-p" style="margin: 0; padding: 0; min-height: 24px; text-indent: 2em"><span class="ne-text" style="color: rgb(36, 41, 47); font-size: 14px">Struction</span><strong><span class="ne-text" style="color: rgb(36, 41, 47); font-size: 14px">:</span></strong></p>

Column | Type | Description
-- | -- | --
help_topic_id | int | topic id
help_keyword_id | int | keyword id

<p id="u8b09edc5" class="ne-p" style="margin: 0; padding: 0; min-height: 24px; text-indent: 2em"><span class="ne-text">Definition:</span></p>
<pre>def_table_schema(
  database_id    = 'OB_MYSQL_SCHEMA_ID',
  table_name    = 'help_relation',
  table_id      = '423',
  table_type = 'SYSTEM_TABLE',
  gm_columns = [],
  rowkey_columns = [
      ('help_topic_id', 'int'),
      ('help_keyword_id', 'int')
  ],
  in_tenant_space = True,
  normal_columns = [],
  columns_with_tenant_id = [],
)      
</pre>
<h2 id="haR8n" style="font-size: 24px; line-height: 32px; margin: 21px 0 5px 0"><span class="ne-text">Create and Initialize </span></h2><p id="u22667b63" class="ne-p" style="margin: 0; padding: 0; min-height: 24px"><span class="ne-text">Since ob supports multi-tenant deployment, every tenant holds its own help tables. For sys-tenant, help tables will be created and initialized during bootstrap. For non-sys tenant, help tables will be created and initialized when the tenant is being created. The way of initialization is to execute the pre-defined sql statement by calling inner sql.</span></p><p id="u33292059" class="ne-p" style="margin: 0; padding: 0; min-height: 24px"><span class="ne-text">How to define the sql statement will be introducted later.</span></p><h2 id="b81Oa" style="font-size: 24px; line-height: 32px; margin: 21px 0 5px 0"><span class="ne-text">Parser the Help Command</span></h2><p id="u676cb95f" class="ne-p" style="margin: 0; padding: 0; min-height: 24px"><span class="ne-text"> 	For parsing help command like &quot;help content&quot;, we need to add corresponding parsing rules in </span><code class="ne-code" style="font-family: SFMono-Regular, Consolas, Liberation Mono, Menlo, Courier, monospace; background-color: rgba(0, 0, 0, 0.06); border: 1px solid rgba(0, 0, 0, 0.08); border-radius: 2px; padding: 0px 2px"><span class="ne-text">./src/sql/parser/sql_parser_mysql_mode.l</span></code><span class="ne-text"> and </span><code class="ne-code" style="font-family: SFMono-Regular, Consolas, Liberation Mono, Menlo, Courier, monospace; background-color: rgba(0, 0, 0, 0.06); border: 1px solid rgba(0, 0, 0, 0.08); border-radius: 2px; padding: 0px 2px"><span class="ne-text">./src/sql/parser/sql_parser_mysql_mode.y</span></code><span class="ne-text">. Then execute </span><code class="ne-code" style="font-family: SFMono-Regular, Consolas, Liberation Mono, Menlo, Courier, monospace; background-color: rgba(0, 0, 0, 0.06); border: 1px solid rgba(0, 0, 0, 0.08); border-radius: 2px; padding: 0px 2px"><span class="ne-text">./src/sql/parser/gen_sql.py</span></code><span class="ne-text">to generate the corresponding c++ files. Help grammar are as follows:</span></p>
<pre>help_stmt:
HELP ident_or_text
{
  $$=$2;
  $$-&gt;type_=T_HELP;
}          
;
ident_or_text:
  NAME_OB     { $$=$1;}
| text_string { $$=$1;}
;
</pre>
<h2 id="Mos43" style="font-size: 24px; line-height: 32px; margin: 21px 0 5px 0"><span class="ne-text">Search and Respond</span></h2><p id="uc925752d" class="ne-p" style="margin: 0; padding: 0; min-height: 24px"><span class="ne-text">According to the content of 'help&quot; command, we need to query different help tables in turn(search topic, search keyword, search topic). We do this by Inner SQL. According to different query results, we construct corresponding select sql statement, which will be parsed and execute later to return the MySQL_RES(the result that client wil be recieve) to client(mysql-client or obclient). Finally, client can analyze the result and output the help contents.</span></p><p id="ub6a9a3ef" class="ne-p" style="margin: 0; padding: 0; min-height: 24px"><span class="ne-text">We do the above operations in the resolver phase.</span></p><h3 id="xvWOc" style="font-size: 20; line-height: 28px; margin: 16px 0 5px 0"><span class="ne-text">seach topic</span></h3><p id="uc130e9bb" class="ne-p" style="margin: 0; padding: 0; min-height: 24px"><span class="ne-text">For exmaple: help &quot;contents&quot;</span></p><p id="u55f5e43f" class="ne-p" style="margin: 0; padding: 0; min-height: 24px"><span class="ne-text">First, we search help topic . We will perform a fuzzy query on the help topic  table by Inner SQl. According to the number of topics queried, wo do different operations as follows:</span></p><ul class="ne-ul" style="margin: 0; padding-left: 23px"><li id="ubec22dbc"><span class="ne-text">topic_num = 0<br /></span><span class="ne-text">It means there has no matching topic, go to the next search stage.</span></li><li id="u67b30af3"><span class="ne-text">topic_num = 1<br /></span><span class="ne-text">It means there has only one matching topic. In this case, we want client to recieve the details of this topic. So the final select sql statement we construct is as follow:<br /></span><code class="ne-code" style="font-family: SFMono-Regular, Consolas, Liberation Mono, Menlo, Courier, monospace; background-color: rgba(0, 0, 0, 0.06); border: 1px solid rgba(0, 0, 0, 0.08); border-radius: 2px; padding: 0px 2px"><span class="ne-text">select name, description, example from mysql.help_topic WHERE name like 'contents'</span></code></li><li id="u13d62e9a"><span class="ne-text">topic_num &gt; 1<br /></span><span class="ne-text">It means there has multiple matching topics. In this case, we want client to recieve the names of these topics. Not only that, we will search the help category table, and get the matching categories to client. After client recieving the result, it will output the topic list and category list(Only name). The final select sql statement we construct is as follow:</span></li></ul>
<pre>select name, &quot;N&quot; as is_it_category from mysql.help_topic 
where name like '%.*s'&quot;
union all
select name, \&quot;Y\&quot; as is_it_category from mysql.help_category 
where name like '%.*s' order by 2 asc , 1 asc&quot;
</pre>
<h3 id="PfIHl" style="font-size: 20; line-height: 28px; margin: 16px 0 5px 0"><span class="ne-text">search keyword</span></h3><p id="ua9741e2e" class="ne-p" style="margin: 0; padding: 0; min-height: 24px"><span class="ne-text">If we didn't get any topic in &quot;seach topic&quot; phase, we search keyword by Inner SQL. We execute the following sql statement to perform a joint query on tables help_keyword , help relation and help topic  to get the matching topics.</span></p><pre data-language="python" id="RO2zy" class="ne-codeblock language-python" style="border: 1px solid #e8e8e8; border-radius: 2px; background: #f9f9f9; padding: 16px; font-size: 13px; color: #595959">select t3.name from mysql.help_keyword t1 
 JOIN mysql.help_relation t2 on (t1.help_keyword_id = t2.help_keyword_id)
 JOIN mysql.help_topic t3 on (t2.help_topic_id = t3.help_topic_id)
 WHERE t1.name like 'contents'.</pre><p id="u7dfb68f4" class="ne-p" style="margin: 0; padding: 0; min-height: 24px"><span class="ne-text"> According to the number of topics queried, wo do different operations as follows(just like 'search topic'):</span></p><ul class="ne-ul" style="margin: 0; padding-left: 23px"><li id="u30a0676a"><span class="ne-text">topic_num = 0<br /></span><span class="ne-text">It means there has no matching topic, go to the next search stage.</span></li><li id="u6d931e6b"><span class="ne-text">topic_num = 1<br /></span><span class="ne-text">It means there has only one matching topic. In this case, we want client to recieve the details of this topic.</span></li><li id="u95075d76"><span class="ne-text">topic_num &gt; 1<br /></span><span class="ne-text">It means there has multiple matching topics. In this case, we want client to recieve the names of these topics. Not only that, we will search the help category table, and get the matching categories to client. After client recieving the result, it will output the topic list and category list(Only name). </span></li></ul><h3 id="tkV4G" style="font-size: 20; line-height: 28px; margin: 16px 0 5px 0"><span class="ne-text">search category</span></h3><p id="ucf237164" class="ne-p" style="margin: 0; padding: 0; min-height: 24px"><span class="ne-text">If we didn't get any topic in &quot;seach topic&quot; and &quot;search keyword&quot; phase, we search category by Inner SQL as follow:</span></p><pre data-language="python" id="HWCQa" class="ne-codeblock language-python" style="border: 1px solid #e8e8e8; border-radius: 2px; background: #f9f9f9; padding: 16px; font-size: 13px; color: #595959">select help_category_id, name from mysql.help_category where name like 'content'</pre><p id="u0d7d9657" class="ne-p" style="margin: 0; padding: 0; min-height: 24px"><span class="ne-text">According to the number of categories queried, wo do different operations as follows(just like 'search topic'): </span></p><ul class="ne-ul" style="margin: 0; padding-left: 23px"><li id="u1a4cadfa"><span class="ne-text">categories_num = 0<br /></span><span class="ne-text">In this case, we trust there has no matching contents, so we construct a result-empty select sql.</span></li><li id="u2b990aca"><span class="ne-text">categories_num = 1<br /></span><span class="ne-text">This situation is more complicated. In this case, In this case, we query help_topic and help_category respectively according to the unique category id we get, and get the topic and subcategories contained in this category.<br /></span><span class="ne-text">In MySQl, the mysql server just return the topics and subcategories(only name), but it may cause bug.(</span><a href="https://bugs.mysql.com/bug.php?id=107997" data-href="https://bugs.mysql.com/bug.php?id=107997" target="_blank" class="ne-link"><span class="ne-text">https://bugs.mysql.com/bug.php?id=107997</span></a><span class="ne-text">)<br /></span><span class="ne-text">So, we take some measures.<br /></span><span class="ne-text">If this is only one matching topic(no matching category), we will return the details of this topic.<br /></span><span class="ne-text">If this is only one matching category(no matching topic), we will return the name of this category.<br /></span><span class="ne-text">Otherwise, we do the same as MySQL.</span></li><li id="u5c2fb200"><span class="ne-text">categories_num &gt; 1<br /></span><span class="ne-text">In this case, we construct a select sql statement to retrun match categories(only names).</span></li></ul><h3 id="dkRcw" style="font-size: 20; line-height: 28px; margin: 16px 0 5px 0"><span class="ne-text">Parse and execute the select sql</span></h3><p id="uec146753" class="ne-p" style="margin: 0; padding: 0; min-height: 24px"><span class="ne-text">After searching topic, keyword and category, we get a select sql statement. Then we do the same operations like &quot;show&quot; command to return the result.</span></p><h1 id="jrnlq" style="font-size: 28px; line-height: 36px; margin: 26px 0 10px 0"><span class="ne-text">How to Contribute</span></h1><p id="ub8d3f783" class="ne-p" style="margin: 0; padding: 0; min-height: 24px"><span class="ne-text">In order to allow more people to enrich the content of the help table. We provide a method. There is a file named </span><code class="ne-code" style="font-family: SFMono-Regular, Consolas, Liberation Mono, Menlo, Courier, monospace; background-color: rgba(0, 0, 0, 0.06); border: 1px solid rgba(0, 0, 0, 0.08); border-radius: 2px; padding: 0px 2px"><span class="ne-text">gen_fill_help_table.py</span></code><span class="ne-text">. You can add category, topic or keyword in category_list, topic_list or keyword_list in a fixed format. After this, execute gen_fill_help_table.py then you need to execute </span><code class="ne-code" style="font-family: SFMono-Regular, Consolas, Liberation Mono, Menlo, Courier, monospace; background-color: rgba(0, 0, 0, 0.06); border: 1px solid rgba(0, 0, 0, 0.08); border-radius: 2px; padding: 0px 2px"><span class="ne-text">obclient -p observer_ip -u root@tenant -P port &lt; ob_fill_help_tables.sql</span></code><span class="ne-text">to update the help tables for any tenant. Newly created tenant will also hold the newest help tables.</span></p><pre data-language="python" id="Klyy1" class="ne-codeblock language-python" style="border: 1px solid #e8e8e8; border-radius: 2px; background: #f9f9f9; padding: 16px; font-size: 13px; color: #595959">category_list = [
    {
		&quot;help_category_id&quot;: &quot;1&quot;,
		&quot;name&quot;: &quot;Contents&quot;,
		&quot;parent_category_id&quot;: &quot;0&quot;,
		&quot;url&quot;: &quot;&quot;
	},
    # add more
]
topic_list = [
    {
		&quot;help_topic_id&quot;: &quot;1&quot;,
		&quot;help_category_id&quot;: &quot;21&quot;,
		&quot;name&quot;: &quot;TINYINT&quot;,
		&quot;description&quot;: &quot;\\n  TINYINT[(M)] [UNSIGNED] [ZEROFILL]\\n\\n &quot;,
		&quot;example&quot;: &quot;&quot;,
		&quot;url&quot;: &quot;&quot;
	},
    # add more
]
keyword_list = [
    {
		&quot;help_keyword_id&quot;: &quot;1&quot;,
		&quot;name&quot;: &quot;TINYINT&quot;,
		&quot;relation_topic&quot;: [1]
	},
    # add more
]</pre><p id="u453feb8e" class="ne-p" style="margin: 0; padding: 0; min-height: 24px"><span class="ne-text">What you can contribute:</span></p><ol class="ne-ol" style="margin: 0; padding-left: 23px"><li id="u5fa426c6"><span class="ne-text">Add a category.<br /></span><span class="ne-text">You need to add a category item in categort_list, then choose a parent-category for it. After that, you can add	some subcategories or topics for it, or not.</span></li><li id="u8fc1326f"><span class="ne-text">Add a topic<br /></span><span class="ne-text">You need to add a topic item in topic_list, and choose a category to hold it.</span></li><li id="ua3609f4d"><span class="ne-text">Add a keyword<br /></span><span class="ne-text">You need to add a keyword item in keyword_list, and bind some topics to it(in the relation_topic filed).</span></li><li id="u9ed5515a"><span class="ne-text">Bind a topic to a keyword<br /></span><span class="ne-text">You only need to add the topic id in the relation_topic filed of a keyword item.</span></li></ol><p id="ud15f3262" class="ne-p" style="margin: 0; padding: 0; min-height: 24px"><span class="ne-text">Just contribute, we will make the correctness check in the script. Make you own help tables.</span></p></div>

**Comment by LINxiansheng** [comment]
+1

**Comment by LINxiansheng** [comment]
Can you add the four virtual tables application in the discussion?

**Reply by Junkrat77** [reply]
Done

**Comment by caifeizhi** [comment]
每张表的数据是如何生成的？

**Reply by Junkrat77** [reply]
在创建租户时,会通过inner sql执行一系列的insert语句来对每张表的数据进行填充,这些sql语句由内部来维护，初稿我会把master中的提供给你
